src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,173 @@
+                lengthArgument);
+
+            var newRoot = root
+                .ReplaceNode(
+                    nodeToFix, 
+                    newNode);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private struct FixComponents
+        {
+            public FixComponents(
+                IOperation stringArgument, 
+                SyntaxNode targetMethod, 
+                ImmutableArray<IArgumentOperation> originalInnerArguments)
+            {
+                StringArgument = stringArgument;
+                TargetMethod = targetMethod;
+                OriginalInnerArguments = originalInnerArguments;
+            }
+
+            public IOperation StringArgument { get; }
+            public SyntaxNode TargetMethod { get; }
+            public ImmutableArray<IArgumentOperation> OriginalInnerArguments { get; }
+        }
+
+        private static async Task<Document> FixCodeTwoParameters(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix = semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            FixComponents fixComponents = GetFixComponents(generator, typedNodeToFix);
+          
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod,
+                fixComponents.StringArgument.Syntax,
+                fixComponents.OriginalInnerArguments[0].Syntax,
+                fixComponents.OriginalInnerArguments[1].Syntax);
+            var newRoot = root.ReplaceNode(
+                nodeToFix,
+                newNode);
+
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private static FixComponents GetFixComponents(
+            SyntaxGenerator generator, 
+            IInvocationOperation typedNodeToFix)
+        {
+            var argumentOperation = typedNodeToFix.Arguments[0].Value;
+            if (argumentOperation is IInvocationOperation stringBuilderAppendInvocationCandidate)
+            {
+                var innerArguments = stringBuilderAppendInvocationCandidate.Arguments;
+                var instance = stringBuilderAppendInvocationCandidate.Instance;
+                
+                var append = generator.MemberAccessExpression(
+                    typedNodeToFix.Instance.Syntax,
+                    generator.IdentifierName(nameof(StringBuilder.Append)));
+                var instanceType = instance.Type;
+
+                return new FixComponents(
+                    stringArgument: instance,
+                    targetMethod: append,
+                    originalInnerArguments: innerArguments);
+            }
+
+            return default(FixComponents);
```

Comment:
i don't see how you properly handle this in the callers.  i see lots of calls that just blindly index in.  but if default is returned here, then things like `.StringArgument.Syntax` in teh caller will end up null-ref'ing.

Reply:
this issue still applies.

Reply:
fixed now by returning instead on the caller.